python/private/py_twine.bzl

```diff
@@ -0,0 +1,106 @@
+
+Simply wraps the tool with a bash script and a runfiles manifest.
+See https://twine.readthedocs.io/
+"""
+
+load(":py_wheel.bzl", "PyWheelInfo")
+
+_attrs = {
+    "twine_bin": attr.label(
+        doc = """\
+A py_binary that runs the twine tool.
+The default value assumes you have `twine` listed in your own requirements.txt, and have run
+`pip_parse` with the default name of `pypi`.
+
+If these don't apply, you might use the `entry_point` helper to supply your own twine binary:
+```starlark
+load("@my_pip_parse_name//:requirements.bzl", "entry_point")
+py_twine(
+    ...
+    twine_bin = entry_point("twine"),
+)
+```
+
+Or of course you can supply a py_binary by some other means which is CLI-compatible with twine.
+
+Currently rules_python doesn't supply twine itself.
+Follow https://github.com/bazelbuild/rules_python/issues/1016
+""",
+        default = "@pypi_twine//:rules_python_wheel_entry_point_twine",
+        executable = True,
+        cfg = "exec",
+    ),
+    "wheel": attr.label(providers = [PyWheelInfo]),
+    "_runfiles_lib": attr.label(default = "@bazel_tools//tools/bash/runfiles"),
+}
+
+# Bash helper function for looking up runfiles.
+# Vendored from
+# https://github.com/bazelbuild/bazel/blob/master/tools/bash/runfiles/runfiles.bash
+BASH_RLOCATION_FUNCTION = r"""
+# --- begin runfiles.bash initialization v2 ---
+set -uo pipefail; f=bazel_tools/tools/bash/runfiles/runfiles.bash
+source "${RUNFILES_DIR:-/dev/null}/$f" 2>/dev/null || \
+source "$(grep -sm1 "^$f " "${RUNFILES_MANIFEST_FILE:-/dev/null}" | cut -f2- -d' ')" 2>/dev/null || \
+source "$0.runfiles/$f" 2>/dev/null || \
+source "$(grep -sm1 "^$f " "$0.runfiles_manifest" | cut -f2- -d' ')" 2>/dev/null || \
+source "$(grep -sm1 "^$f " "$0.exe.runfiles_manifest" | cut -f2- -d' ')" 2>/dev/null || \
+{ echo>&2 "ERROR: cannot find $f"; exit 1; }; f=; set -e
+# --- end runfiles.bash initialization v2 ---
+"""
+
+# Copied from https://github.com/aspect-build/bazel-lib/blob/main/lib/private/paths.bzl
+# to avoid a dependency from bazelbuild -> aspect-build
+def _to_manifest_path(ctx, file):
+    """The runfiles manifest entry path for a file.
+
+    This is the full runfiles path of a file including its workspace name as
+    the first segment. We refert to it as the manifest path as it is the path
+    flavor that is used for in the runfiles MANIFEST file.
+    We must avoid using non-normalized paths (workspace/../other_workspace/path)
+    in order to locate entries by their key.
+    Args:
+        ctx: starlark rule execution context
+        file: a File object
+    Returns:
+        The runfiles manifest entry path for a file
+    """
+
+    if file.short_path.startswith("../"):
+        return file.short_path[3:]
+    else:
+        return ctx.workspace_name + "/" + file.short_path
+
+_exec_tmpl = """\
+#!/usr/bin/env bash
+{rlocation}
+tmp=$(mktemp -d)
+# The namefile is just a file with one line, containing the real filename for the wheel.
```

Comment:
Why cant the produced file have this name?

Reply:
Because the analysis phase wants to predict the output files and the status files can only be read during action execution.

Reply:
What part of the file name isn't knowable ahead of time? We're producing a final name such as: `runfiles-0.1.2-py39-none-any.whl` right? The first part is defined, the last 3 are part of configuration info available to the rule implementation. The 2nd part is variable, but is something that the CI knows (because it gets it from the git tag, right?) and can pass along on the command line.

Reply:
https://github.com/bazelbuild/rules_python/issues/741

Reply:
"pass along on the command line" - I'm not sure what you mean by this, we have an action under `bazel build` that outputs the file

Reply:
The user already has stamping setup for their Bazel workspace, when they say `--stamp` and maybe `--workspace_status_command=./some/script.sh` that produces values in the `bazel-out/stable-status.txt` and `bazel-out/volatile-status.txt` files. That's the API contract for stamping. I think it's not ideal for us to say "by the way, if some of your artifacts are wheel files, then stamping isn't enough and you need to change how you run Bazel in your CI/CD pipeline"

Reply:
Well, I don't agree, because stamping is metadata about the whole build, not specific artifacts. But lets take the convo to the issue instead of the PR, where stamping and wheels are the dedicated topic, as it doesn't need to block this PR.